File: solutions/lua/sum-of-multiples/1/sum-of-multiples.lua
local function sum_keys(o)
	local acc = 0
	for k, _ in pairs(o) do
		acc = acc + k
	end
	return acc
end

return function(factors)
	local res = {}
	function res.to(limit)
    local muls = {}

		for _, fac in ipairs(factors) do
			if fac ~= 0 then
				local i = 0
				while i*fac < limit do
					muls[i*fac] = true
					i = i + 1
				end
			end
		end

		return sum_keys(muls)
	end
	return res
end

File: solutions/ocaml/sum-of-multiples/1/sum_of_multiples.ml
open Base

let sum factors limit =
  List.fold_left factors ~init:(Set.empty (module Int)) ~f:(fun set fac ->
    if fac <> 0 then
      List.range ~stride:fac 0 limit |> List.fold_left ~init:set ~f:Set.add
    else set
  )
  |> Set.sum (module Int) ~f:(fun x -> x)
